common/raw_hashtable_metadata_group_benchmark.cpp

```diff
@@ -0,0 +1,233 @@
+static auto BuildBenchMetadata() -> llvm::ArrayRef<BenchMetadata> {
+  static uint8_t metadata_storage[GroupSize][BenchSize * GroupSize];
+  static uint8_t bytes_storage[GroupSize][BenchSize];
+  static BenchMetadata bm_storage[GroupSize];
+  absl::BitGen gen;
+  for (auto [bm_index, bm] : llvm::enumerate(bm_storage)) {
+    int match_count = bm_index + 1;
+
+    for (ssize_t g_index : llvm::seq<ssize_t>(0, BenchSize)) {
+      // Start by filling the group with random bytes.
+      auto group_bytes = llvm::MutableArrayRef(
+          &metadata_storage[bm_index][g_index * GroupSize], GroupSize);
+      for (uint8_t& b : group_bytes) {
+        b = absl::Uniform<uint8_t>(gen) | MetadataGroup::PresentMask;
+      }
+
+      // Now we need up to `match_count` random indices into the group where
+      // we'll put a matching byte.
+      std::array<ssize_t, GroupSize> group_indices;
+      std::iota(group_indices.begin(), group_indices.end(), 0);
+      std::shuffle(group_indices.begin(), group_indices.end(), gen);
+
+      // Now cause the first match index to have the desired value.
+      ssize_t match_index = *group_indices.begin();
+      uint8_t& match_b = group_bytes[match_index];
+      switch (Kind) {
+        case BenchKind::Random: {
+          // Already a random value, but we need to  ensure it isn't one that
+          // repeats elsewhere in the group.
+          while (llvm::count(group_bytes, match_b) > 1) {
+            match_b = absl::Uniform<uint8_t>(gen) | MetadataGroup::PresentMask;
+          }
+          // Store this as the byte to search for in this group, but without the
+          // present bit off to simulate where we start when using a 7-bit tag
+          // from a hash.
+          bytes_storage[bm_index][g_index] =
+              match_b & ~MetadataGroup::PresentMask;
+          break;
+        }
+        case BenchKind::Empty: {
+          match_b = MetadataGroup::Empty;
+          break;
+        }
+        case BenchKind::Deleted: {
+          match_b = MetadataGroup::Deleted;
+          break;
+        }
+      }
+
+      // Replicate the match byte in each of the other matching indices.
+      for (ssize_t m_index : llvm::ArrayRef(group_indices)
+                                 .drop_front()
+                                 .take_front(match_count - 1)) {
+        group_bytes[m_index] = match_b;
+      }
+    }
+
+    // Now that the storage is set up, record these in our struct.
+    bm.metadata = metadata_storage[bm_index];
+    if constexpr (Kind == BenchKind::Random) {
+      bm.bytes = bytes_storage[bm_index];
+    }
+  }
+  return bm_storage;
+}
+
+template <BenchKind Kind>
+// NOLINTNEXTLINE(google-readability-casting): False positive clang-tidy bug.
+const auto bench_metadata = BuildBenchMetadata<Kind>();
+
+template <BenchKind Kind, typename GroupT = MetadataGroup>
+static void BM_LoadMatch(benchmark::State& s) {
+  BenchMetadata bm = bench_metadata<Kind>[0];
+  ssize_t nonce_data[GroupSize];
```

Comment:
Comment what this is trying to accomplish (something about defeating optimizations and creating dependencies between iterations so as to make this more of a latency benchmark than a throughput benchmark?)

(similarly below)

Reply:
Copious comments added throughout this code, and some renaming to try to clarify as well.